source/common/http/http2/conn_pool.cc

```diff
@@ -15,176 +15,260 @@ namespace Envoy {
 namespace Http {
 namespace Http2 {
 
+using Upstream::ConnectionRequestPolicy;
+
 ConnPoolImpl::ConnPoolImpl(Event::Dispatcher& dispatcher, Upstream::HostConstSharedPtr host,
                            Upstream::ResourcePriority priority,
                            const Network::ConnectionSocket::OptionsSharedPtr& options)
     : ConnPoolImplBase(std::move(host), std::move(priority)), dispatcher_(dispatcher),
       socket_options_(options) {}
 
-ConnPoolImpl::~ConnPoolImpl() {
-  if (primary_client_) {
-    primary_client_->client_->close();
-  }
-
-  if (draining_client_) {
-    draining_client_->client_->close();
-  }
+void ConnPoolImpl::applyToEachClient(std::list<ActiveClientPtr>& client_list,
+                                     const std::function<void(const ActiveClientPtr&)>& fn) {
+  for_each(client_list.begin(), client_list.end(), fn);
+}
 
-  // Make sure all clients are destroyed before we are destroyed.
+ConnPoolImpl::~ConnPoolImpl() {
+  drainConnections();
```

Comment:
All connections need to be closed here in this case like happened previously, not just put in the draining state.

Reply:
It happens in the method `drainConnections`. It in turn calls `checkForDrained` which then calls `close` on all `closable` clients. I thought once we move all the clients to `drain` state, it implies that we dont accept any new requests and just wait for existing requests to finish .